File: trunk/FireflowSuite/engine/src/org/fireflow/engine/kenelextensions/TransitionInstanceExtension.java
/**
 * Copyright 2007-2008 非也
 * All rights reserved. 
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation。
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses. *
 */
package org.fireflow.engine.kenelextensions;

import java.util.Map;
import org.fireflow.engine.IRuntimeContextAware;
import org.fireflow.engine.RuntimeContext;
import org.fireflow.engine.condition.ConditionConstant;
import org.fireflow.kenel.ITransitionInstance;
import org.fireflow.kenel.KenelException;
import org.fireflow.engine.condition.IConditionResolver;
import org.fireflow.kenel.event.ITransitionInstanceEventListener;
import org.fireflow.kenel.event.TransitionInstanceEvent;
import org.fireflow.kenel.impl.TransitionInstance;
import org.fireflow.kenel.plugin.IKenelExtension;

/**
 *
 * @author chennieyun
 */
public class TransitionInstanceExtension implements IKenelExtension,
        ITransitionInstanceEventListener, IRuntimeContextAware {

    protected RuntimeContext rtCtx = null;

    public void setRuntimeContext(RuntimeContext ctx) {
        this.rtCtx = ctx;
    }

    public RuntimeContext getRuntimeContext() {
        return this.rtCtx;
    }

    /**
     * 执行分支判断策略，即设置token的alive属性 首先，如果this.alive==false,则所有的token的Alive属性皆为false
     * 然后，如果在nexttransitionList中有值，则直接执行列表中的tansition
     * 否则，通过计算Transition的表达式来确定下一个transition,
     * 
     * @param transInst
     * @return
     */
    private boolean determineTheAliveOfToken(Map vars, String condition) {
//        System.out.println("Inside SynchronizerInstance.DeterminTheAliveOfToken():: joinPoint.getAlive =" + joinPoint.getAlive());
//        if (!joinPoint.getAlive()) {
//            return false;
//        }
//        Set<String> nextTransitionInstanceNames = joinPoint.getAppointedTransitionNames();
//        if (nextTransitionInstanceNames.size() > 0) {
//            Iterator nextTransNamesIterator = nextTransitionInstanceNames.iterator();
//            while (nextTransNamesIterator.hasNext()) {
//                String transName = (String) nextTransNamesIterator.next();
//                if (transName.equals(transInst.getTransition().getName())) {
//                    return true;
//                }
//            }
//        }
//
//        if (transInst.getTransition().getCondition() == null || transInst.getTransition().getCondition().trim().equals("")) {
//            return true;
//        }

        // TODO通过计算transition上的表达式来确定alive的值

        IConditionResolver elResolver = this.rtCtx.getConditionResolver();
        Boolean b = elResolver.resolveBooleanExpression(vars, condition);

        return b;
    }

    public String getExtentionTargetName() {
        return TransitionInstance.Extension_Target_Name;
    }

    public String getExtentionPointName() {
        return TransitionInstance.Extension_Point_TransitionInstanceEventListener;
    }

    public void onTransitionInstanceEventFired(TransitionInstanceEvent e) throws KenelException {
        if (e.getEventType() == TransitionInstanceEvent.ON_TAKING_THE_TOKEN) {
            if (!e.getToken().isAlive()) {
                return;//如果token是dead状态，表明synchronizer的joinpoint是dead状态，不需要重新计算。
            //计算token的alive值
            }
            ITransitionInstance transInst = (ITransitionInstance) e.getSource();
            String condition = transInst.getTransition().getCondition();

            //1、如果没有转移条件，默认为true
            if (condition == null || condition.trim().equals("")) {
                e.getToken().setAlive(true);
                return;
            }
            //2、default类型的不需要计算其alive值，该值由synchronizer决定
            if (condition.trim().equals(ConditionConstant.DEFAULT)) {
                return;
            }

            //3、计算EL表达式
            boolean alive = determineTheAliveOfToken(e.getToken().getProcessInstance().getProcessInstanceVariables(), condition);
            e.getToken().setAlive(alive);
        }
    }
}

File: trunk/dot-net_version/FireWorkflow.Net/FireWorkflow.Net/Engine/Kernelextensions/TransitionInstanceExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using FireWorkflow.Net.Engine;
using FireWorkflow.Net.Engine.Impl;
using FireWorkflow.Net.Engine.Condition;
using FireWorkflow.Net.Kernel;
using FireWorkflow.Net.Kernel.Event;
using FireWorkflow.Net.Kernel.Impl;
using FireWorkflow.Net.Kernel.Plugin;
using FireWorkflow.Net.Model;
using FireWorkflow.Net.Model.Net;

namespace FireWorkflow.Net.Engine.Kernelextensions
{
    public class TransitionInstanceExtension : IKernelExtension,
        IEdgeInstanceEventListener, IRuntimeContextAware
    {

        protected RuntimeContext rtCtx = null;

        public void setRuntimeContext(RuntimeContext ctx)
        {
            this.rtCtx = ctx;
        }

        public RuntimeContext getRuntimeContext()
        {
            return this.rtCtx;
        }

        /// <summary>
        /// 执行分支判断策略，即设置token的alive属性 首先，如果this.alive==false,则所有的token的Alive属性皆为false
        /// 然后，如果在nexttransitionList中有值，则直接执行列表中的tansition
        /// 否则，通过计算Transition的表达式来确定下一个transition,
        /// </summary>
        /// <param name="vars"></param>
        /// <param name="condition"></param>
        /// <returns></returns>
        private Boolean determineTheAliveOfToken(Dictionary<String, Object> vars, String condition)
        {
            //        System.out.println("Inside SynchronizerInstance.DeterminTheAliveOfToken():: joinPoint.getAlive =" + joinPoint.getAlive());
            //        if (!joinPoint.getAlive()) {
            //            return false;
            //        }
            //        Set<String> nextTransitionInstanceNames = joinPoint.getAppointedTransitionNames();
            //        if (nextTransitionInstanceNames.Count > 0) {
            //            Iterator nextTransNamesIterator = nextTransitionInstanceNames.iterator();
            //            while (nextTransNamesIterator.hasNext()) {
            //                String transName = (String) nextTransNamesIterator.next();
            //                if (transName.Equals(transInst.getTransition().Name)) {
            //                    return true;
            //                }
            //            }
            //        }
            //
            //        if (transInst.getTransition().getCondition() == null || transInst.getTransition().getCondition().trim().Equals("")) {
            //            return true;
            //        }

            // TODO通过计算transition上的表达式来确定alive的值

            IConditionResolver elResolver = this.rtCtx.getConditionResolver();
            Boolean b = elResolver.resolveBooleanExpression(vars, condition);

            return b;
        }

        public void calculateTheAliveValue(IToken token, String condition)
        {

            if (!token.isAlive())
            {
                return;//如果token是dead状态，表明synchronizer的joinpoint是dead状态，不需要重新计算。
            }

            //1、如果没有转移条件，默认为true
            if (condition == null || condition.Trim().Equals(""))
            {
                token.setAlive(true);
                return;
            }
            //2、default类型的不需要计算其alive值，该值由synchronizer决定
            if (condition.Trim().Equals(ConditionConstant.DEFAULT))
            {
                return;
            }

            //3、计算EL表达式
            try
            {
                Boolean alive = determineTheAliveOfToken(token.getProcessInstance().getProcessInstanceVariables(), condition);
                token.setAlive(alive);
            }
            catch (Exception ex)
            {
                throw new EngineException(token.getProcessInstanceId(), token.getProcessInstance().getWorkflowProcess(), token.getNodeId(), ex.Message);
            }

        }

        public String getExtentionTargetName()
        {
            return TransitionInstance.Extension_Target_Name;
        }

        public String getExtentionPointName()
        {
            return TransitionInstance.Extension_Point_TransitionInstanceEventListener;
        }

        public void onEdgeInstanceEventFired(EdgeInstanceEvent e)
        {
            if (e.getEventType() == EdgeInstanceEvent.ON_TAKING_THE_TOKEN)
            {
                IToken token = e.getToken();
                ITransitionInstance transInst = (ITransitionInstance)e.getSource();
                String condition = transInst.getTransition().Condition;
                calculateTheAliveValue(token, condition);

                if (rtCtx.isEnableTrace() && token.isAlive())
                {
                    Transition transition = transInst.getTransition();
                    IWFElement fromNode = transition.FromNode;
                    int minorNumber = 1;
                    if (fromNode is Activity)
                    {
                        minorNumber = 2;
                    }
                    else
                    {
                        minorNumber = 1;
                    }

                    ProcessInstanceTrace trace = new ProcessInstanceTrace();
                    trace.setProcessInstanceId(e.getToken().getProcessInstanceId());
                    trace.setStepNumber(e.getToken().getStepNumber());
                    trace.setType(ProcessInstanceTrace.TRANSITION_TYPE);
                    trace.setFromNodeId(transInst.getTransition().FromNode.Id);
                    trace.setToNodeId(transInst.getTransition().ToNode.Id);
                    trace.setEdgeId(transInst.getTransition().Id);
                    trace.setMinorNumber(minorNumber);
                    rtCtx.getPersistenceService().saveOrUpdateProcessInstanceTrace(trace);
                }
            }

        }
    }
}
